tests/test_backup.py

```diff
@@ -367,6 +367,145 @@ def test_delete_backup(self, mock_available_backups, tmpdir, caplog):
+        mock_stat,
+        fsync,
+    ):
+        """
+        Test that the _set_backup_sizes method correctly sets the backup sizes
+        and optionally performs fsync.
+        """
+        # Set up the mocks
+        backup_manager = build_backup_manager()
+        mock_stat.reset_mock()
+        mock_backup_info = Mock()
+
+        # Mock os.walk to return a predefined directory structure
+        mock_walk.return_value = [
+            ("/root", ["dir1", "dir2"], ["file1.txt"]),
+            ("/root/dir1", [], ["file2.txt"]),
+            ("/root/dir2", ["subdir"], []),
+            ("/root/dir2/subdir", [], ["file3.txt"]),
+        ]
+
+        # Define the mock return values for os.stat
+        def mock_stat_return_value(backup):
+            return_values = {
+                "/root/file1.txt": {
+                    "size": 1024,
+                    "nlink": 3,
+                },
+                "/root/dir1/file2.txt": {
+                    "size": 2048,
+                    "nlink": 2,
+                },
+                "/root/dir2/subdir/file3.txt": {
+                    "size": 4096,
+                    "nlink": 1,
+                },
+            }
+            return Mock(
+                st_size=return_values[backup]["size"],
+                st_nlink=return_values[backup]["nlink"],
+            )
+
+        mock_stat.side_effect = mock_stat_return_value
+
+        # Define the mock return values for fsync_file
+        def mock_fsync_file_return_value(file_path):
+            return mock_stat_return_value(file_path)
+
+        mock_fsync_file.side_effect = mock_fsync_file_return_value
+
+        # Call the method under test
+        backup_manager._set_backup_sizes(mock_backup_info, fsync)
+
+        # Assertions for both with and without fsync cases
+        mock_walk.assert_called_once_with(
+            mock_backup_info.get_basebackup_directory.return_value,
+        )
+        assert mock_backup_info.set_attribute.call_count == 2
+        mock_backup_info.save.assert_called_once()
+
+        # Assertions when called with fsync
+        if fsync:
+            mock_stat.assert_not_called()
+            assert mock_fsync_dir.call_count == 4
+            mock_fsync_dir.assert_has_calls(
+                [
+                    call("/root"),
+                    call("/root/dir1"),
+                    call("/root/dir2"),
+                    call("/root/dir2/subdir"),
+                ]
+            )
+            assert mock_fsync_file.call_count == 3
+            mock_fsync_file.assert_has_calls(
+                [
+                    call("/root/file1.txt"),
+                    call("/root/dir1/file2.txt"),
+                    call("/root/dir2/subdir/file3.txt"),
+                ]
+            )
+            mock_backup_info.set_attribute.assert_has_calls(
+                [
+                    call("size", 7168),
+                    call("deduplicated_size", 4096),
+                ]
+            )
```

Comment:
I think this can be moved out of the `if-else` branch because it's a common assertion.